### PR TITLE
upgrade: Run drift along with version check

### PR DESCRIPTION
**Pull request by efritz**
This adds a pre-upgrade/downgrade check for schema drift. Users can bypass this (but it's not strongly advertised). This should reduce failed upgrade/downgrades due to mismatched schema definitions.

## Test plan

Ran locally with a purposefully busted schema.

### CODE REVIEW

File: internal/database/migration/cliutil/downgrade.go [hunk 1]
```diff
@@ -16,6 +16,7 @@ func Downgrade(
 	runnerFactory RunnerFactoryWithSchemas,
 	outFactory OutputFactory,
 	registerMigrators func(storeFactory migrations.StoreFactory) oobmigration.RegisterMigratorsFunc,
+	expectedSchemaFactories ...ExpectedSchemaFactory,
```
Comment (tbliu98) [inline internal/database/migration/cliutil/downgrade.go:19]: Are there plans to do this for downgrades as well?

File: internal/database/migration/cliutil/multiversion.go [hunk 2]
```diff
@@ -110,9 +111,11 @@ func runMigration(
 	privilegedMode runner.PrivilegedMode,
 	privilegedHash string,
 	skipVersionCheck bool,
+	skipDriftCheck,
```
Comment (pjlast) [inline internal/database/migration/cliutil/multiversion.go:114]:
```suggestion
	skipDriftCheck bool,
```

File: internal/database/migration/cliutil/multiversion.go [hunk 3]
```diff
@@ -275,3 +284,72 @@ func setServiceVersion(ctx context.Context, r Runner, version oobmigration.Versi
...
+			empty, err := isEmptySchema(ctx, r, schemaName)
+			if err != nil {
+				return err
+			}
+			if empty {
+				continue
+			}
+
+			drift = true
+			out.WriteLine(output.Linef(output.EmojiSuccess, output.StyleFailure, "Schema drift detected for %s", schemaName))
```
Comment (pjlast) [inline internal/database/migration/cliutil/multiversion.go:322]: EmojiFailure perhaps🤔? feels weird to get a ✅ for something that's wrong

File: internal/database/migration/cliutil/multiversion.go [hunk 4]
```diff
@@ -275,3 +284,72 @@ func setServiceVersion(ctx context.Context, r Runner, version oobmigration.Versi
...
+		noopOutput := output.NewOutput(&buf, output.OutputOpts{})
+
+		if err := compareByFactories(schemaName, plan.from.GitTag(), schema, noopOutput, expectedSchemaFactories); err != nil {
+			schemasWithDrift = append(schemasWithDrift, schemaName)
+		}
+	}
+
+	drift := false
+
+	if len(schemasWithDrift) > 0 {
```
Comment (pjlast) [inline internal/database/migration/cliutil/multiversion.go:311]: Why the `len` check before the `range`? Think if the list is empty then the range would just be skipped

File: internal/database/migration/cliutil/multiversion.go (hunk 3 above)
Comment (efritz) [reply internal/database/migration/cliutil/multiversion.go:322]: Whoops, yeah good catch.

File: internal/database/migration/cliutil/multiversion.go (hunk 4 above)
Comment (efritz) [reply internal/database/migration/cliutil/multiversion.go:311]: I think this is an artifact of writing this in several ways when I had a side-effect at the end of the loop when the condition was true. Will simplify.

**Review by eseliger — APPROVED** [review]
rubber stamp

**Pull request merged**